velox/functions/sparksql/CharVarcharUtils.h

```diff
@@ -0,0 +1,103 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <string>
+#include "../Macros.h"
+#include "../lib/string/StringImpl.h"
+
+namespace facebook::velox::functions::sparksql {
+
+/// Trims trailing ASCII space characters (0x20) from ``inputStr``
+/// to ensure its length does not exceed the specified Unicode string length
+/// ``limit`` in characters. Throws an exception if the string still exceeds
+/// `limit` after trimming.
+FOLLY_ALWAYS_INLINE void trimTrailingSpaces(
+    exec::StringWriter& output,
+    StringView inputStr,
+    int32_t numChars,
+    int32_t limit) {
+  const auto numTailSpacesToTrim = numChars - limit;
+  VELOX_USER_CHECK_GT(numTailSpacesToTrim, 0);
+
+  auto curPos = inputStr.end() - 1;
+  const auto trimTo = inputStr.end() - numTailSpacesToTrim;
+
+  while (curPos >= trimTo && stringImpl::isAsciiSpace(*curPos)) {
+    curPos--;
+  }
+  // Get the length of the trimmed string in characters.
+  const auto trimmedSize = numChars - std::distance(curPos + 1, inputStr.end());
+
+  VELOX_USER_CHECK_LE(
+      trimmedSize,
+      limit,
+      "Exceeds char/varchar type length limitation: {}",
+      limit);
+  output.setNoCopy(
+      StringView(inputStr.data(), std::distance(inputStr.begin(), curPos + 1)));
+}
+
+/// Ensures that ``input`` fit within the specified length ``limit`` in
+/// characters. If the length of ``input`` exceeds the limit, trailing spaces
+/// are trimmed to fit within the limit. If the length of ``input`` is less than
+/// or equal to
+/// ``limit``, it is returned as-is. Throws exception if the trimmed string
+/// still exceeds ``limit`` or if ``limit`` is negative. This function will trim
+/// at most (length of ``input`` - ``limit``) space characters (ASCII 32) from
+/// the end of ``input``.
+template <typename T>
+struct VarcharTypeWriteSideCheckFunction {
```

Comment:
Move the function to a separate file like 'VarcharTypeWriteSideCheck.h', and keep the helper function in the Utils.